File: apps/vscode/src/heartbeat.ts
import * as vscode from 'vscode';
import * as config from './config';
import * as detector from './detector';
import { HeartbeatQueue } from './queue';

export interface HeartbeatPayload {
    file: string;
    project: string;
    language: string;
    editor: 'vscode';
    branch?: string;
    repoUrl?: string;
    websiteUrl?: string;
    lastCommitAt?: number;
    os?: string;
    machine?: string;
    lines?: number;
    cursorLine?: number;
    timezone?: string;
    keystrokes: number;
    time: number;
}

const HEARTBEAT_INTERVAL_MS = 30 * 1000;

export class HeartbeatService {
    private queue = new HeartbeatQueue();
    private hasShownInvalidKeyWarning = false;

    // Characters inserted since the last heartbeat was sent.
    private keystrokeCount = 0;

    // The document to report on next tick, and whether anything has
    // happened since the last send. Every editor event just updates
    // these — it never triggers a network call directly. Only the
    // periodic tick() below decides when an actual heartbeat goes out,
    // so no matter how many times you switch tabs, save, or type, at
    // most one heartbeat fires per 30-second interval.
    private pendingDocument: vscode.TextDocument | null = null;
    private hasPendingActivity = false;

    private timer: ReturnType<typeof setInterval> | undefined;

    /** Called from onDidChangeTextDocument with characters inserted in that edit. */
    recordKeystrokes(charsInserted: number): void {
        if (charsInserted <= 0) return;
        this.keystrokeCount += charsInserted;
    }

    /**
     * Called from every relevant editor event (typing, switching files,
     * saving, gaining focus). Just records what's being worked on right
     * now — sending is entirely decided by the timer in start().
     */
    noteActivity(document: vscode.TextDocument): void {
        if (!config.isEnabled()) return;
        if (!detector.shouldTrack(document)) return;

        this.pendingDocument = document;
        this.hasPendingActivity = true;
    }

    /** Starts the periodic flush loop. Call once from activate(). */
    start(): void {
        this.timer = setInterval(() => {
            void this.tick();
        }, HEARTBEAT_INTERVAL_MS);
    }

    dispose(): void {
        if (this.timer) clearInterval(this.timer);
    }

    private async tick(): Promise<void> {
        if (!config.hasApiKey()) return;
        if (!this.hasPendingActivity || !this.pendingDocument) return;

        const document = this.pendingDocument;
        this.hasPendingActivity = false;

        const payload = await this.buildPayload(document);
        this.keystrokeCount = 0; // captured into payload above, start counting fresh
        await this.send(payload);
    }

    private async buildPayload(document: vscode.TextDocument): Promise<HeartbeatPayload> {
        const editor = vscode.window.activeTextEditor;

        const useGitRootProjectName = config.useGitRootProjectName();
        const metadata = await detector.detectProjectMetadata(document, useGitRootProjectName);

        return {
            file: document.fileName,
            project: await detector.detectProjectName(document, useGitRootProjectName),
            language: document.languageId,
            editor: 'vscode',
            branch: await detector.detectBranch(),
            repoUrl: metadata.repoUrl,
            websiteUrl: metadata.websiteUrl,
            lastCommitAt: metadata.lastCommitAt,
            os: detector.detectOS(),
            machine: detector.detectMachine(),
            lines: document.lineCount,
            cursorLine: editor ? editor.selection.active.line + 1 : undefined,
            timezone: detector.detectTimezone(),
            keystrokes: this.keystrokeCount,
            time: Date.now(),
        };
    }

    private async send(payload: HeartbeatPayload): Promise<void> {
        const apiKey = config.getApiKey();
        const apiUrl = config.getApiUrl();

        try {
            const res = await fetch(`${apiUrl}/api/heartbeat`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    Authorization: `Bearer ${apiKey}`,
                },
                body: JSON.stringify(payload),
            });

            if (res.status === 401) {
                this.notifyInvalidKey();
                return;
            }

            if (res.ok) {
                // Since we're online right now, also try to clear
                // out anything stuck in the offline queue.
                await this.queue.flush(apiKey, apiUrl);
            } else {
                this.queue.enqueue(payload);
            }
        } catch {
            // Network error (offline, DNS failure, etc) — queue it
            // silently and try again later. Never bother the user.
            this.queue.enqueue(payload);
        }
    }

    private notifyInvalidKey(): void {
        if (this.hasShownInvalidKeyWarning) return;
        this.hasShownInvalidKeyWarning = true;
        vscode.window.showWarningMessage(
            'Seismic: Invalid API key. Run "Seismic: Set API Key" to update it.',
        );
    }

    /**
     * Called periodically in the background to retry any
     * heartbeats that failed to send earlier.
     */
    async flushQueue(): Promise<void> {
        await this.queue.flush(config.getApiKey(), config.getApiUrl());
    }
}

File: apps/nvim/lua/seismic/heartbeat.lua
local M = {}

local config = require("seismic.config")
local detector = require("seismic.detector")
local queue = require("seismic.queue")

local HEARTBEAT_INTERVAL_SECONDS = 30

local last_heartbeat_time = 0
local last_file = ""
local has_shown_invalid_key_warning = false
local keystroke_count = 0

function M.record_keystroke()
	keystroke_count = keystroke_count + 1
end

local function build_payload(bufnr)
	local lines = vim.api.nvim_buf_line_count(bufnr)
	local cursor_line = vim.api.nvim_win_get_cursor(0)[1]

	return {
		file = vim.api.nvim_buf_get_name(bufnr),
		project = detector.detect_project(config.use_git_root_project_name()),
		language = detector.language_id(bufnr),
		editor = "neovim",
		branch = detector.detect_branch(),
		os = detector.detect_os(),
		machine = detector.detect_machine(),
		lines = lines,
		keystrokes = keystroke_count,
		cursorLine = cursor_line,
		timezone = detector.detect_timezone(),
		time = os.time() * 1000,
	}
end

local function notify_invalid_key()
	if has_shown_invalid_key_warning then
		return
	end
	has_shown_invalid_key_warning = true
	vim.notify("Seismic: Invalid API key. Run :SeismicSetApiKey to update it.", vim.log.levels.WARN)
end

local function send(payload)
	local api_key = config.get_api_key()
	local api_url = config.get_api_url()
	local body = vim.json.encode(payload)

	vim.system({
		"curl", "-s", "-o", "/dev/null", "-w", "%{http_code}",
		"-X", "POST",
		"-H", "Content-Type: application/json",
		"-H", "Authorization: Bearer " .. api_key,
		"-d", body,
		api_url .. "/api/heartbeat",
	}, { text = true }, function(result)
		local code = result.stdout

		if code == "401" then
			notify_invalid_key()
			return
		end

		if code:match("^2%d%d") then
			queue.flush(api_key, api_url)
		else
			queue.enqueue(payload)
		end
	end)
end

function M.handle_activity(bufnr, forced)
	if not config.is_enabled() then
		return
	end
	if not config.has_api_key() then
		return
	end
	if not detector.should_track(bufnr) then
		return
	end

	local now = os.time()
	local filename = vim.api.nvim_buf_get_name(bufnr)
	local file_changed = filename ~= last_file

	if not forced and not file_changed and (now - last_heartbeat_time) < HEARTBEAT_INTERVAL_SECONDS then
		return
	end

	last_heartbeat_time = now
	last_file = filename

	local payload = build_payload(bufnr)
	keystroke_count = 0
	send(payload)
end

function M.flush_queue()
	queue.flush(config.get_api_key(), config.get_api_url())
end

return M
